prov/efa/test/efa_unit_test_common.c

```diff
@@ -180,3 +182,61 @@ void efa_unit_test_eager_msgrtm_pkt_construct(struct rxr_pkt_entry *pkt_entry, s
 	assert_int_equal(*connid, attr->connid);
 	pkt_entry->pkt_size = sizeof(base_hdr) + sizeof(opt_connid_hdr);
 }
+
+/**
+ * @brief Construct RXR_HANDSHAKE_PKT
+ *	The function will include the optional connid/host id headers if and only if
+ *	attr->connid/host id are non-zero.
+ *
+ * @param[in,out] pkt_entry Packet entry. Must be non-NULL.
+ * @param[in] attr Packet attributes.
+ */
+void efa_unit_test_handshake_pkt_construct(struct rxr_pkt_entry *pkt_entry, struct efa_unit_test_handshake_pkt_attr *attr)
```

Comment:
Why do we need this function?

Cannot we just use `rxr_pkt_init_handshake`?

Reply:
We cannot use `rxr_pkt_init_handshake` - that is part of the logic I am testing against so we cannot directly use it.

The current testing flow starts from the receiver of the handshake packet:
1. Local endpoint receives a handshake packet (constructed from this function) from a remote peer.
2. The endpoint processes the packet, updates the peer host id and posts back a handshake to the peer, which calls `rxr_pkt_init_handshake`

The test asserts on 1) the peer host id is updated according to the received handshake packet, and 2) the local endpoint correctly sends back a handshake packet to the peer that optionally includes local host id.

Reply:
Could you elaborate you concerns?

I actually think this is where unit test shines. With mocks we can verify fine-grained aspects of code behavior. Especially for host id - which is a private endpoint attribute underneath the `fi_*` APIs - we can test its value at different points of endpoint lifecycle.

I'm not sure how we would test the same aspects in integration test - do we need to inspect the endpoint memory structure, or filter EFA packet to find the host id? The only quick way I can think of is through logging - but that is not a good approach. Please let me know.

Reply:
I see. thanks!